helix-view/src/editor.rs

```diff
@@ -178,6 +179,44 @@ pub struct Config {
     pub indent_guides: IndentGuidesConfig,
     /// Whether to color modes with different colors. Defaults to `false`.
     pub color_modes: bool,
+    pub soft_wrap: SoftWrap,
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+#[serde(default, rename_all = "kebab-case", deny_unknown_fields)]
+pub struct SoftWrap {
+    /// Soft wrap lines that exceed viewport width. Default to off
+    pub enable: bool,
+    /// Maximum space that softwrapping may leave free at the end of the line when perfomring softwrapping
+    /// This space is used to wrap text at word boundries. If that is not possible within this limit
+    /// the word is simply split at the end of the line.
+    ///
+    /// This is automatically hardlimited to a quarter of the viewport to ensure correct display on small views.
+    ///
+    /// Default to 5
+    pub max_wrap: u16,
+    /// Maximum number of indentation that can be carried over from the previous line when softwrapping.
+    /// If a line is indenten further then this limit it is rendered at the start of the viewport instead.
+    ///
+    /// This is automatically hardlimited to a quarter of the viewport to ensure correct display on small views.
+    ///
+    /// Default to 40
+    pub max_indent_retain: u16,
+    /// Extra spaces inserted before rendeirng softwrapped lines.
+    ///
+    /// Default to 2
+    pub wrap_indent: u16,
+}
+
+impl Default for SoftWrap {
```

Comment:
Turns out if you only enable softwrap but don't set the other values then it will cause a panic, as default won't "create" the other values, maybe some of the other subtable configs have the way to add the partial defaults

Reply:
Could you post your config and the backtrace? Locally `:set softwrap.enable` true was enough for me

Reply:
I added it above but it happens when you just have and no other values in the subtable
```toml
[config.soft-wrap]
enable = true
```

Reply:
I only have

``` toml

[editor.soft-wrap]
enable=true
```

in my config and it works without issue. It's really odd that it would be causing problems for you.
Maybe a backtrace will help

Reply:
sorry let me get some more details

Reply:
I tried both `:set soft-wrap.enable true` and
```toml
[editor.soft-wrap]
enable=true
```
without changing anything else in my current config and it works without issues.

Reply:
ok this is very specific and I doesn't have to do with what I intetially thought, that was premature sorry.

to get the panic open run helix from this branch in the helix repository and then open the picker and search `r` the full command is ``<space>fr`` then you will get this backtrace

```
thread 'main' panicked at 'called `Option::unwrap()` on a `None` value', helix-term/src/ui/document.rs:158:40
stack backtrace:
   0: rust_begin_unwind
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:143:14
   2: core::panicking::panic
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/panicking.rs:48:5
   3: core::option::Option<T>::unwrap
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/option.rs:755:21
   4: helix_term::ui::document::render_text
             at ./helix-term/src/ui/document.rs:158:26
   5: helix_term::ui::document::render_document
             at ./helix-term/src/ui/document.rs:103:5
   6: <helix_term::ui::picker::FilePicker<T> as helix_term::compositor::Component>::render
             at ./helix-term/src/ui/picker.rs:304:13
   7: <helix_term::ui::overlay::Overlay<T> as helix_term::compositor::Component>::render
             at ./helix-term/src/ui/overlay.rs:48:9
   8: helix_term::compositor::Compositor::render
             at ./helix-term/src/compositor.rs:167:13
   9: helix_term::application::Application::render::{{closure}}
             at ./helix-term/src/application.rs:311:9
  10: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  11: helix_term::application::Application::handle_terminal_events::{{closure}}
             at ./helix-term/src/application.rs:634:26
  12: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  13: helix_term::application::Application::event_loop_until_idle::{{closure}}
             at ./helix-term/src/application.rs:347:55
  14: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  15: helix_term::application::Application::event_loop::{{closure}}
             at ./helix-term/src/application.rs:326:57
  16: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  17: helix_term::application::Application::run::{{closure}}
             at ./helix-term/src/application.rs:1071:38
  18: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  19: hx::main_impl::{{closure}}
             at ./helix-term/src/main.rs:156:53
  20: <core::future::from_generator::GenFuture<T> as core::future::future::Future>::poll
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/future/mod.rs:91:19
  21: tokio::runtime::park::CachedParkThread::block_on::{{closure}}
             at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/park.rs:283:63
  22: tokio::runtime::coop::with_budget
             at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/coop.rs:102:5
  23: tokio::runtime::coop::budget
             at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/coop.rs:68:5
  24: tokio::runtime::park::CachedParkThread::block_on
             at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/park.rs:283:31
  25: tokio::runtime::context::BlockingRegionGuard::block_on
             at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/context.rs:295:13
  26: tokio::runtime::scheduler::multi_thread::MultiThread::block_on
             at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/scheduler/multi_thread/mod.rs:66:9
  27: tokio::runtime::runtime::Runtime::block_on
             at /home/gaby/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/runtime/runtime.rs:284:45
  28: hx::main_impl
             at ./helix-term/src/main.rs:158:5
  29: hx::main
             at ./helix-term/src/main.rs:38:21
  30: core::ops::function::FnOnce::call_once
             at /rustc/fe5b13d681f25ee6474be29d748c65adcd91f69e/library/core/src/ops/function.rs:227:5
```

Reply:
Thanks for reproducing! I actually ran into this edgecase earlier but was unable to reproduce. It only shows up if the picker preview is visible (I sually test with a very small window to actually get softwrapping in all files so its usually not openend while testing).
The crash is actually entirely unrelated to softwrapping but rather seems like some kind of minor oversight while refactoring the text rendering. I will look into a fix now.

Reply:
the file that the picker should show right before it panics is ``rustfmt.toml`` that seems like an empty file that might be why this is happening

Reply:
You are right that is the cause.
I have already found it fixed the issue.
The problem was that I am skipping empty highlight ranges in the highlight iterator (required for correctness) but I was unwrapping the first syntax highlight range. For an empty file there is only a single syntax highlight range because the only visible range is empty.
The bug was a bit easy to miss as it only occurs in the picker (when a file is opened the cursor always causes the EOF to be highlighted so the highlight range is never truly empty)